realm/realm-library/src/main/cpp/io_realm_internal_sync_OsSubscription.cpp

```diff
@@ -0,0 +1,120 @@
+
+#include <results.hpp>
+#include <sync/partial_sync.hpp>
+
+using namespace realm;
+using namespace realm::jni_util;
+using namespace realm::_impl;
+
+typedef ObservableCollectionWrapper<Results> ResultsWrapper;
+
+static void finalize_subscription(jlong ptr);
+
+static void finalize_subscription(jlong ptr)
+{
+    TR_ENTER_PTR(ptr);
+    delete reinterpret_cast<SubscriptionWrapper*>(ptr);
+}
+
+JNIEXPORT jlong JNICALL Java_io_realm_internal_sync_OsSubscription_nativeCreate(JNIEnv* env, jclass, jlong results_ptr, jstring j_subscription_name)
+{
+    TR_ENTER()
+    try {
+        const auto results = reinterpret_cast<ResultsWrapper*>(results_ptr);
+        JStringAccessor subscription_name(env, j_subscription_name);
+        auto key = subscription_name.is_null_or_empty() ? util::none : util::Optional<std::string>(subscription_name);
+        auto subscription = partial_sync::subscribe(results->collection(), key);
+        auto wrapper = new SubscriptionWrapper(std::move(subscription));
+        return reinterpret_cast<jlong>(wrapper);
+    }
+    CATCH_STD()
+    return reinterpret_cast<jlong>(nullptr);
+}
+
+
+JNIEXPORT jlong JNICALL Java_io_realm_internal_sync_OsSubscription_nativeGetFinalizerPtr(JNIEnv*, jclass)
+{
+    TR_ENTER()
+    return reinterpret_cast<jlong>(&finalize_subscription);
+}
+
+JNIEXPORT void JNICALL Java_io_realm_internal_sync_OsSubscription_nativeStartListening(JNIEnv* env, jobject object, jlong native_ptr)
+{
+    TR_ENTER()
+    try {
+        auto wrapper = reinterpret_cast<SubscriptionWrapper*>(native_ptr);
+        wrapper->start_listening(env, object);
+    }
+    CATCH_STD()
+}
+
+JNIEXPORT void JNICALL Java_io_realm_internal_sync_OsSubscription_nativeStopListening(JNIEnv* env, jobject, jlong native_ptr)
+{
+    TR_ENTER()
+    try {
+        auto wrapper = reinterpret_cast<SubscriptionWrapper*>(native_ptr);
+        wrapper->stop_listening();
+    }
+    CATCH_STD()
+}
+
+JNIEXPORT jint JNICALL Java_io_realm_internal_sync_OsSubscription_nativeGetState(JNIEnv* env, jclass, jlong native_ptr)
+{
+    TR_ENTER()
+    try {
+        auto wrapper = reinterpret_cast<SubscriptionWrapper*>(native_ptr);
+        return static_cast<jint>(wrapper->subscription().state());
+    }
+    CATCH_STD()
+    return 0;
+}
+
+JNIEXPORT jobject JNICALL Java_io_realm_internal_sync_OsSubscription_nativeGetError(JNIEnv* env, jclass, jlong native_ptr)
+{
+    TR_ENTER()
+    try {
+        auto wrapper = reinterpret_cast<SubscriptionWrapper*>(native_ptr);
+        auto err = wrapper->subscription().error();
+        if (err) {
+            std::string error_message = "";
+            if (err) {
+                try {
+                    std::rethrow_exception(err);
+                }
+                catch (const std::exception &e) {
+                    error_message = e.what();
+                }
+            }
+
+            static JavaClass illegal_argument_class(env, "java/lang/IllegalArgumentException");
```

Comment:
Why `IllegalArgumentException`? are you assuming the error is related to the subscription parameters? we deprecated the dynamic query for partial sync AFAIK, so this exception cannot be something the user did wrong (besides the naming perhaps)

Reply:
Some queries are still disallowed, like backlink queries, but the most common error is probably re-using a subscription name for a different query. 

Theoretically, a few very rare exceptions might happen here that are not considered `IllegalArgument`,  but I don't even know where to begin to filter for those, so I just left everything as `IllegalArgument` instead of the generic `RealmException`. I think that is a sensible trade-off?